File: Algorithm/Heuristic/atcoder/TravelingSalesmanProblem/A46/A46.php
<?php

fscanf(STDIN, "%d", $n);
$cities = [];

for ($i = 0; $i < $n; $i++) {
    fscanf(STDIN, "%d %d", $x, $y);
    $cities[] = [$x, $y];
}

// 都市間距離（ユークリッド距離の2乗）
function dist($a, $b, $cities) {
    $dx = $cities[$a][0] - $cities[$b][0];
    $dy = $cities[$a][1] - $cities[$b][1];
    return $dx * $dx + $dy * $dy;
}

// 最近傍法による初期ツアー生成
function greedyTour($n, $cities) {
    $visited = array_fill(0, $n, false);
    $tour = [0];
    $visited[0] = true;

    while (count($tour) < $n) {
        $last = end($tour);
        $next = -1;
        $minDist = PHP_INT_MAX;

        for ($i = 0; $i < $n; $i++) {
            if (!$visited[$i]) {
                $d = dist($last, $i, $cities);
                if ($d < $minDist) {
                    $minDist = $d;
                    $next = $i;
                }
            }
        }
        $tour[] = $next;
        $visited[$next] = true;
    }
    return $tour;
}

// 2-opt 法
function twoOpt($tour, $cities) {
    $n = count($tour);
    $improved = true;

    while ($improved) {
        $improved = false;

        for ($i = 1; $i < $n - 1; $i++) {
            for ($j = $i + 1; $j < $n; $j++) {
                $a = $tour[$i - 1];
                $b = $tour[$i];
                $c = $tour[$j];
                $d = $tour[($j + 1) % $n];

                $before = dist($a, $b, $cities) + dist($c, $d, $cities);
                $after  = dist($a, $c, $cities) + dist($b, $d, $cities);

                if ($after < $before) {
                    $tour = array_merge(
                        array_slice($tour, 0, $i),
                        array_reverse(array_slice($tour, $i, $j - $i + 1)),
                        array_slice($tour, $j + 1)
                    );
                    $improved = true;
                }
            }
        }
    }

    return $tour;
}

// 実行
$tour = greedyTour($n, $cities);
$tour = twoOpt($tour, $cities);

// 出力（1-indexed）
foreach ($tour as $t) {
    echo ($t + 1) . "\n";
}
echo ($tour[0] + 1) . "\n";

File: Algorithm/Heuristic/atcoder/TravelingSalesmanProblem/A46/A46.py
import sys
import math
import time

sys.setrecursionlimit(10000)

# --- 入力 ---
N = int(sys.stdin.readline())
cities = [tuple(map(int, sys.stdin.readline().split())) for _ in range(N)]

# --- 距離関数（平方ユークリッド距離） ---
def dist(i, j):
    dx = cities[i][0] - cities[j][0]
    dy = cities[i][1] - cities[j][1]
    return dx * dx + dy * dy

# --- 初期解：貪欲法（最近傍法） ---
def greedy_tour():
    visited = [False] * N
    tour = [0]  # 都市1（インデックス0）からスタート
    visited[0] = True
    for _ in range(N - 1):
        last = tour[-1]
        next_city = min(
            ((i, dist(last, i)) for i in range(N) if not visited[i]),
            key=lambda x: x[1]
        )[0]
        tour.append(next_city)
        visited[next_city] = True
    return tour

# --- 2-opt法 ---
def two_opt(tour):
    improved = True
    while improved:
        improved = False
        for i in range(1, N - 2):
            for j in range(i + 1, N - 1):
                a, b = tour[i - 1], tour[i]
                c, d = tour[j], tour[j + 1]
                before = dist(a, b) + dist(c, d)
                after = dist(a, c) + dist(b, d)
                if after < before:
                    tour[i:j + 1] = reversed(tour[i:j + 1])
                    improved = True
        # optional: break early if time is tight
    return tour

# --- 実行 ---
start_time = time.time()

tour = greedy_tour()
tour = two_opt(tour)

# --- 出力（都市番号を1-indexedで） ---
for city in tour:
    print(city + 1)
print(tour[0] + 1)  # 最後に都市1へ戻る

# --- optional: debug info ---
# print("Total squared distance:", sum(dist(tour[i], tour[(i+1)%N]) for i in range(N)))
# print("Time used:", time.time() - start_time)
